design/baremetalhost-v1alpha2.md

```diff
@@ -0,0 +1,236 @@
+when applying rules.
+
+The best approach is to use an `int32` field.
+
+### Converting Existing Data
+
+It is possible to [use a web
+hook](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/#configure-customresourcedefinition-to-use-conversion-webhooks)
+to convert resources from their storage format to a specific API
+format. We do not want users to continue to consume the `v1alpha1`
+format, however, because we want to be able to drop support
+quickly. Therefore, we should not write a conversion web hook.
+
+Instead, we should add startup code to the baremetal-operator to read
+all host resources in their current format and write them (with type
+conversion but no other changes) to the new storage format. See option
+2
+[here](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/#upgrade-existing-objects-to-a-new-stored-version)
+for more discussion.
+
+### Dropping Hardware Profile Field
+
+Now that we have support for more verbose root device hints and we
+have the `hardware-classification-controller`, we no longer need the
+hardware profile fields. We can remove them as part of the
+backwards-incompatible changes as we move to the new API version.
+
+### Risks and Mitigations
+
+We do not know all of the ways that consumers of the API are using the
+CPU speed values currently being reported. Given that most kubernetes
+tools appear to assume that floating point types are not to be used in
+APIs, it is likely that most consumers are either ignoring the value
+or using it in simple ways (displaying it, logging it, etc.) where
+loss of precision from the type conversion to a float is not
+significant.
+
+### Work Items
+
+- Release `baremetal-operator` version `0.1.0` with existing API.
+- Add new `v1alpha2` API
+- Change the `ClockSpeed` type to `int32` in `v1alpha2`.
+- Remove hardware profile from spec and status.
+- Remove the state machine states for setting the hardware profile.
+- Set `v1alpha2` as the storage version
+- Add conversion code to the `baremetal-operator`
+- Update the inspection code to no longer convert the value given by
+  Ironic Inspector before storing it.
+- Release `baremetal-operator` version `0.2.0` with code that can
+  upgrade existing resources.
+- Update the module references in `cluster-api-provider-metal3` and
+  `hardware-classification-controller`.
+- Drop support for `v1alpha1`
+  - Remove the API code
+  - Remove the conversion code
+  - Remove the `operator-sdk` code from the `tools` directory
+- Release `baremetal-operator` version `0.3.0`.
+- Update the module references in `cluster-api-provider-metal3` and
+  `hardware-classification-controller`.
+
+### Dependencies
+
+The prerequisite work was done as part of
+[#434](https://github.com/metal3-io/baremetal-operator/issues/434).
+
+### Test Plan
+
+- Test updating the `baremetal-operator` on an existing cluster with
```

Comment:
we should probably include an e2e test that drives an upgrade with all pieces, including CAPI / CAPM3 and HWCC. What I am after is to test the integration around the new feature during the upgrade, to make sure that CAPM3 will be able to use the BMH v1a2 after upgrade from v1a1, not only with new v1a2 BMHs.
This might be more complex to put in place, as we'll need in that case to do the setup through clusterctl and then proceed with the upgrade.

Reply:
Given the nature of the changes planned, it seems unlikely that we'll have any issue. Putting an upgrade test in place does make sense, though. What sort of test scenarios should we include? Just upgrading and ensuring the new controllers don't put resources into an error state isn't enough. We probably need to reprovision a host so we can see that everything actually works. Maybe we can build that test before starting the rest of this work, based on existing CAPBM releases. What do you think?

Reply:
IMO, the tests, if added, should probably interact with BMO through consumers, for example updating a machine deployment from CAPI or modifying a HWCC. There's some similar work on-going in metal3-dev-env for upgrade verifications. THis could be synced, because the delta does not seem that big.